src/helpers/buffers/fsv.rs

```diff
@@ -38,62 +28,93 @@ impl<const N: usize> FixedSizeByteVec<N> {
+        debug_assert!(
+            !self.added.replace(i, true),
+            "Duplicate send for index {index} on channel {channel:?}"
+        );
         self.data[offset].copy_from_slice(elem);
     }
 
-    /// Returns `true` if record at the given index exists.
-    pub fn added(&self, index: usize) -> bool {
-        self.added[index]
+    /// Return any gap ahead of the first missing value.
+    #[cfg(any(test, debug_assertions))]
+    pub fn missing(&self) -> Range<usize> {
+        let start = self.end - self.capacity;
+        let i = self.end % self.capacity;
+        let mut absent = self.added[i..].leading_zeros();
+        if i + absent == self.capacity {
+            absent += self.added[..i].leading_zeros();
+        }
+        if absent == self.capacity {
+            start..start
+        } else {
+            start..(start + absent)
+        }
     }
 
     /// Takes a block of elements from the beginning of the vector, or `None` if
     /// fewer than `min_count` elements have been inserted at the start of the buffer.
     pub fn take(&mut self, min_count: usize) -> Option<Vec<u8>> {
-        let contiguous = self.added.leading_ones();
-
-        if contiguous < min_count {
+        // Find the relative index we're starting at.
+        let i = self.end % self.capacity;
+
+        // Find how many elements we can return (`tail + wrap`).
+        let tail = self.added[i..].leading_ones();
+        let wrap = if tail + i == self.capacity {
+            self.added[..i].leading_ones()
+        } else {
+            0
+        };
+
+        if tail + wrap < min_count {
             return None;
         }
-        self.added.drain(..contiguous).for_each(drop);
-        let r = self.data.drain(..contiguous * N).collect();
-        self.taken += contiguous;
-
-        // clear out last `contiguous` elements in the buffer
-        self.added.resize(self.added.len() + contiguous, false);
-        self.data.resize(self.data.len() + contiguous * N, 0);
-
-        Some(r)
-    }
 
-    /// Returns total number of elements evicted from this buffer since the creation.
-    pub fn taken(&self) -> usize {
-        self.taken
-    }
-
-    /// returns the maximum number of elements this vector can hold.
-    pub fn capacity(&self) -> usize {
-        self.data.capacity() / N
+        // Move `self.end` marker, clear the values in `self.added`, and
+        // return a copy of that part of `self.data` that matters.
+        self.end += tail + wrap;
+        self.added[i..(i + tail)].fill(false);
+        if wrap > 0 {
+            self.added[..wrap].fill(false);
+            let mut buf = Vec::with_capacity((tail + wrap) * N);
+            buf.extend_from_slice(&self.data[(i * N)..]);
+            buf.extend_from_slice(&self.data[..(wrap * N)]);
+            Some(buf)
+        } else {
+            Some(self.data[(i * N)..((i + tail) * N)].to_owned())
+        }
     }
 }
 
 #[cfg(all(test, not(feature = "shuttle")))]
 mod tests {
```

Comment:
I would feel a bit more confident if we added a randomized test that uses capacity values picked arbitrarily and tests that take/insert work correctly. bugs in circular buffers are often subtle. Having said that, I don't see anything wrong with this implementation